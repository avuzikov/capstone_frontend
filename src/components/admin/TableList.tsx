import React from 'react';
import TableCard from './TableCard';

const tables = [
  { name: 'Users', link: '/admin/tables/users' },
  { name: 'Jobs', link: '/admin/tables/jobs' },
  { name: 'Applications', link: '/admin/tables/applications' },
];

console.log('Rendering TableList with tables:', tables);

const TableList = () => {
  return (
    <div className="flex flex-col gap-3">
      {tables.map(table => (
        <TableCard key={table.name} link={table.link} name={table.name} />
      ))}
    </div>
  );
};

export default TableList;
