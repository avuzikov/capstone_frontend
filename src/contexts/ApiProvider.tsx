class ApiProvider {
    private static apiUrl = '/users'; // Define the base URL for the API
  
    // Método para manejar el login
    static async login(email: string, password: string): Promise<{ token: string }> {
      try {
        const response = await fetch(`${this.apiUrl}/login`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ email, password }),
        });
  
        const data = await response.json();
  
        if (!response.ok) {
          throw new Error(data.message || 'Login failed');
        }
  
        const authToken = response.headers.get('Authorization');
        if (!authToken) {
          throw new Error('Token not found');
        }
  
        // Guardar el token en una cookie o en localStorage
        //document.cookie = `authToken=${authToken}; path=/`;
        return { token: authToken };
      } catch (error: any) {
        console.error('Login error:', error.message);
        throw error;
      }
    }
  
    // Método para manejar el logout
    static logout(): void {
      // Limpiar la cookie o localStorage
      document.cookie = 'authToken=; Max-Age=0; path=/'; // Eliminar el token
      console.log('Logged out');
    }
  
    // Puedes agregar más métodos para otras operaciones con la API
    static async getUserData(): Promise<any> {
      try {
        const response = await fetch(`${this.apiUrl}/profile`, {
          method: 'GET',
          headers: {
            'Authorization': `Bearer ${this.getAuthToken()}`, // Usa el token para acceder a recursos protegidos
          },
        });
  
        const data = await response.json();
  
        if (!response.ok) {
          throw new Error(data.message || 'Failed to fetch user data');
        }
  
        return data;
      } catch (error: any) {
        console.error('Fetch error:', error.message);
        throw error;
      }
    }
  
    // Método helper para obtener el token de las cookies
    private static getAuthToken(): string | null {
      const name = 'authToken=';
      const decodedCookie = decodeURIComponent(document.cookie);
      const ca = decodedCookie.split(';');
      for (let i = 0; i < ca.length; i++) {
        let c = ca[i].trim();
        if (c.indexOf(name) === 0) {
          return c.substring(name.length, c.length);
        }
      }
      return null;
    }
  }
  
  export default ApiProvider;