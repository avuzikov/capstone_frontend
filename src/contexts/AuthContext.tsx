import React, { createContext, useContext, useState, useEffect } from "react";

interface AuthContextType {
  token: string | null;
  role: string | null;
  login: (email: string, password: string) => Promise<void>;
  logout: () => void;
  setData: (token: string, role: string) => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [token, setToken] = useState<string | null>(
    localStorage.getItem("authToken")
  );
  const [role, setRole] = useState<string | null>(
    localStorage.getItem("userRole")
  );

  useEffect(() => {
    if (token) {
      localStorage.setItem("authToken", token);
    } else {
      localStorage.removeItem("authToken");
    }
  }, [token]);

  useEffect(() => {
    if (role) {
      localStorage.setItem("userRole", role);
    } else {
      localStorage.removeItem("userRole");
    }
  }, [role]);

  const setData = (token: string, role: string) => {
    setToken(token);
    setRole(role);
  };

  const login = async (email: string, password: string) => {
    const response = await fetch("/users/login", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ email, password }),
    });

    if (response.ok) {
      const data = await response.json();
      setToken(data.token);
      setRole(data.role);
    } else {
      throw new Error("Login failed");
    }
  };

  const logout = () => {
    setToken(null);
    setRole(null);
  };

  return (
    <AuthContext.Provider value={{ token, role, login, logout, setData }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
