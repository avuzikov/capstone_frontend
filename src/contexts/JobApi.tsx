// src\contexts\JobApi.tsx

export const fetchJobs = async (page: number, items: number, query = '', token: string | null) => {
  const url =
    query.length > 0
      ? `/api/job/search?value=${query}&page=${page}&items=${items}`
      : `/api/job?page=${page}&items=${items}`;

  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to fetch jobs: Unauthorized');
  }

  const data = await response.json();
  return data;
};

export const fecthApplications = async (page: number, items: number, token: string | null) => {
  const response = await fetch(`/api/application?page=${page}&items=${items}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to fetch applications: Unauthorized');
  }

  const data = await response.json();
  console.log(data);
  return data;
};
