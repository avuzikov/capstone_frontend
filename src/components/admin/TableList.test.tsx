// src\components\admin\TableList.test.tsx

import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TableList from './TableList';
import { useAuth } from '../../contexts/AuthContext';

// Mock the useAuth hook
jest.mock('../../contexts/AuthContext', () => ({
  useAuth: jest.fn(),
}));

// Mock the TableCard component
jest.mock('./TableCard.tsx', () => {
  return ({ link, name }: { link: string; name: string }) => (
    <div data-testid="table-card">
      <a href={link}>{name}</a>
    </div>
  );
});

describe('TableList Component', () => {
  const mockToken = 'mock-token';
  const mockTables = [
    { id: '1', name: 'Users', link: '/admin/tables/users' },
    { id: '2', name: 'Jobs', link: '/admin/tables/jobs' },
    { id: '3', name: 'Applications', link: '/admin/tables/jobs' },
  ];

  beforeEach(() => {
    (useAuth as jest.Mock).mockReturnValue({ token: mockToken });
    global.fetch = jest.fn();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('displays table cards on successful fetch', async () => {
    (fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => mockTables,
    });

    render(<TableList />);

    await waitFor(() => {
      const tableCards = screen.getAllByTestId('table-card');
      expect(tableCards).toHaveLength(mockTables.length);
      mockTables.forEach(table => {
        expect(screen.getByText(table.name)).toBeInTheDocument();
      });
    });
  });

  test('h1 elements have correct text content', async () => {
    (fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => mockTables,
    });

    render(<TableList />);

    //TODO: Look over this test
    await waitFor(() => {
      mockTables.forEach(table => {
        const h1Element = screen.getByText(table.name);
        console.log('H1:', h1Element.innerHTML);
        console.log('TAG Name', h1Element);
        expect(h1Element.innerHTML).toBe(table.name);
      });
    });
  });
});
