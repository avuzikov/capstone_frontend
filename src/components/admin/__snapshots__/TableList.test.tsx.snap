// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-3"
  />
</DocumentFragment>
`;
